torch/sparse/__init__.py

```diff
@@ -201,3 +201,154 @@ def log_softmax(input: Tensor, dim: int, dtype: Optional[DType] = None) -> Tenso
+
+When all elements of :attr:`input` along the given dimension
+:attr:`dim` are ignored, the corresponding element of the output
+tensor will have undefined value: it may or may not correspond to the
+identity value of {operation name} operation; the choice may
+correspond to the value that leads to the most efficient storage of
+:attr:`output` tensor.
+
+The shapes of the :attr:`mask` tensor and the :attr:`input` tensor
+don't need to match, but they must be :ref:`broadcastable
+<broadcasting-semantics>`.
+
+Args:
+    input (Tensor): the input tensor
+    dim (int): a dimension along which {function name} will be
+      computed.
+
+Keyword args:
+    keepdim (bool, optional): whether the output tensor has
+      :attr:`dim` retained or not. Default: False.
+    dtype (:class:`torch.dtype`, optional): the desired data type
+      of returned tensor.  If specified, the input tensor is
+      casted to :attr:`dtype` before the operation is
+      performed. Default: None
+    mask (:class:`torch.Tensor`, optional): the boolean tensor
+      containing the binary mask of validity of input tensor
+      elements.
+      Default: ``torch.ones(input.shape, dtype=torch.bool)``.''',
+        masked_reduction_example='''\
+Example::
+
+    >>> input = torch.tensor([[-3, -2, -1], [0, 1, 2]])
+    >>> input
+    tensor([[-3, -2, -1],
+            [ 0,  1,  2]])
+    >>> mask = torch.tensor([[True, False, True], [False, False, False]])
+    >>> mask
+    tensor([[ True, False,  True],
+            [False, False, False]])''',
+        masked_normalization_args='''\
+    TBD
+''')
+
+    # Apply function name info to docstring templates:
+    templates = dict(
+        (k, v.format_map(
+            {'function_name': func.__name__,
+             'function name': func.__name__.replace('_', ' '),
+             'operation_name': '_'.join(func.__name__.split('_')[1:]),
+             'operation name': ' '.join(func.__name__.split('_')[1:])}
+        )) for k, v in docstring_templates.items())
+
+    # Apply docstring templates to function doctring:
+    func.__doc__ = func.__doc__.format_map(templates)
+
+    # Expose function as public symbol
+    __all__.append(func.__name__)
+
+    return func
+
+
+@_apply_docstring_templates
+def masked_sum(input: Tensor,
+               mask: Optional[Tensor] = None,
+               dim: Optional[int] = None,
+               *,
+               keepdim: Optional[bool] = False,
+               dtype: Optional[DType] = None,
+               ) -> Tensor:
+    """
+{masked_reduction_signature}
+
+{masked_reduction_descr}
+
+{masked_reduction_args}
+
+{masked_reduction_example}
+    >>> torch.sparse.masked_sum(input, mask, 1)
+    tensor([-4,  0])
+    """
+    # TODO: What follows is a reference implementation of masked sum
+    # that is to be replaced with an optimized one.
+    if mask is None:
+        input_mask = torch.as_tensor(input)
+    else:
+        input_mask = torch.as_tensor(input) * torch.as_tensor(mask)
```

Comment:
You actually have to use masked_fill because
```
>>> torch.tensor([float('inf')]) * torch.tensor([False])
tensor([nan])
>>> torch.tensor([float('inf')]).masked_fill(~torch.tensor([False]), 0)
tensor([0.])
```
will cause NaNs in the reduction.

Also, sounds like a good test case to feed NaNs and Infs into masked out elements and masking sure they don't propagate

Reply:
I'm also wondering whether it's necessary to explicitly cast mask to the dtype of the input Tensor. What if the input is bool or int8?

Reply:
@cpuhrsch This implementation is not final nor correct as noted in the TODO comment and what you observed. This implementation should be ignored at this stage of the PR, in fact, the PR was created to discuss the testing approach as it uses OpInfo support in an untypical way (reference is pytorch implementation rather than numpy one).

Reply:
Added tests with nan/inf and replaced input_mask computation to use `where`.